### ISSUE TITLE
Infinite restart_syscall loop on rt_sigsuspend debugging gdb itself

### ISSUE THREAD

**Issue by pwaller** [mozilla/rr#2714]
Ubuntu 20.04. 
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
rr version 5.3.0

#### Command to reproduce
```
rr record gdb -batch -ex r --args echo
```

#### Background

I'm trying to debug gdb itself, to explore some broken behaviour in gdb.

Unfortunately, if I run rr on gdb, it looks like it loops infinitely on rt_sigsuspend.

I have to kill rr. It seems to end up with a trace, and if I do `rr dump` it shows this:
<details>
<summary>Click to expand `rr dump` output</summary>

```
[normal-looking process startup stuff]
{
  real_time:5117.070782 global_time:2188, event:`SYSCALL: rt_sigsuspend' (state:EXITING_SYSCALL) tid:91136, ticks:17622000
rax:0xfffffffffffffdfe rbx:0x82 rcx:0xffffffffffffffff rdx:0x1 rsi:0x8 rdi:0x55897d9eaf00 rbp:0x681fffa0 rsp:0x681ffe10 r8:0x7b5f637 r9:0x7f r10:0x0 r11:
0x246 r12:0x0 r13:0x0 r14:0x55897f5c3370 r15:0x7fff48735200 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x82 fs_base
:0x7f69249f6600 gs_base:0x0
}
{
  real_time:5117.070861 global_time:2189, event:`SYSCALL: restart_syscall' (state:ENTERING_SYSCALL) tid:91136, ticks:17622000
rax:0xffffffffffffffda rbx:0x82 rcx:0xffffffffffffffff rdx:0x1 rsi:0x8 rdi:0x55897d9eaf00 rbp:0x681fffa0 rsp:0x681ffe10 r8:0x7b5f637 r9:0x7f r10:0x0 r11:
0x246 r12:0x0 r13:0x0 r14:0x55897f5c3370 r15:0x7fff48735200 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x82 fs_base
:0x7f69249f6600 gs_base:0x0
}
{
  real_time:5122.617261 global_time:2190, event:`SYSCALL: restart_syscall' (state:EXITING_SYSCALL) tid:91136, ticks:17622000
rax:0xfffffffffffffdfe rbx:0x82 rcx:0xffffffffffffffff rdx:0x1 rsi:0x8 rdi:0x55897d9eaf00 rbp:0x681fffa0 rsp:0x681ffe10 r8:0x7b5f637 r9:0x7f r10:0x0 r11:0x246 r12:0x0 r13:0x0 r14:0x55897f5c3370 r15:0x7fff48735200 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x82 fs_base:0x7f69249f6600 gs_base:0x0
}
[infinite repeats of enter/exit restart_syscall events here]
```

</details>

ticks does not advance. RR consumes 80% of the CPU and GDB consumes 20% of the cpu. `perf trace -p $(pgrep gdb)` shows only a high rate of rt_sigsuspend calls.

**Comment by rocallahan** [comment]
Have you got exact steps to reproduce?

**Comment by pwaller** [comment]
Goodness, apologies! :)

`rr record gdb -batch -ex r --args echo`

**Comment by rocallahan** [comment]
Modern gdbs embed `guile`, which uses `libgc` i.e. the BDW conservative GC, which does various tricky things that don't work with rr's syscall buffering. For example syscall buffering switches the stack while we emulate a syscall, and libgc doesn't understand that. I think there are other issues too but I haven't fully investigated since currently the effort required to support libgc can't be justified. So, when recording libgc/guile/gdb you'll have to `rr record -n` and accept the performance penalty.

This doesn't fix the problem you reported though.

**Comment by rocallahan** [comment]
The problem you reported is something I actually wrote about in March!
https://robert.ocallahan.org/2020/03/debugging-gdb-using-rr-ptrace-emulation.html

TL;DR
```
maint set catch-demangler-crashes off
```

**Comment by pwaller** [comment]
Oh fantastic, thanks for the quick response! I should have searched harder :). It does indeed work on the master branch.

From the post:

> I suspect that (with KPTI especially) doing 2-3 syscalls per symbol demangle (sigprocmask is also called) hurts gdb performance even without rr, so ideally someone would fix that.

... ! For what it's worth I just had a look, but the number of calls I see with strace to rt_* is slightly higher, but it's not very large in my typical use of gdb on a large C++ project with lots of symbols.

**Comment by rocallahan** [comment]
gdb may do lazy symbol demangling in some situations. I'm not sure.

If you find that gdb demangles more symbols when run under rr, that'd be interesting, or if you find that gdb is doing a lot of rt_sigaction syscalls not associated with symbol demangling, that'd also be interesting, otherwise there isn't anything to do.
